File: src/main/scala/net/machinemuse/powersuits/block/ModelLuxCapatitor.scala
//package net.machinemuse.powersuits.block
//
//import com.google.common.base.Function
//import com.google.common.collect.ImmutableMap
//import net.machinemuse.numina.geometry.Colour
//import net.machinemuse.powersuits.common.ModularPowersuits
//import net.minecraft.client.Minecraft
//import net.minecraft.client.renderer.block.model.IBakedModel
//import net.minecraft.client.renderer.block.model.ModelRotation
//import net.minecraft.client.renderer.texture.TextureAtlasSprite
//import net.minecraft.client.renderer.vertex.DefaultVertexFormats
//import net.minecraft.util.EnumFacing
//import net.minecraft.util.ResourceLocation
//import net.minecraftforge.client.model.ModelLoaderRegistry
//import net.minecraftforge.client.model.obj.OBJLoader
//import net.minecraftforge.client.model.obj.OBJModel
//import net.minecraftforge.common.model.TRSRTransformation
//import net.minecraftforge.fml.relauncher.Side
//import net.minecraftforge.fml.relauncher.SideOnly
//import javax.vecmath.Matrix4f
//import javax.vecmath.Vector4f
//import java.util.HashMap
//import java.util.Map
//
///**
//  * Created by leon on 9/14/16.
//  */
//object ModelLuxCapatitor {
//  private val modelLocation: ResourceLocation = new ResourceLocation(ModularPowersuits.MODID.toLowerCase, "models/block/luxcapacitor.obj")
//  private val materialName: String = "Material__3.003" // material of the lens
//  private val customData: util.Map[String, String] = new util.HashMap[String, String]() {}
//}
//
//class ModelLuxCapatitor {
//  private val registeredModels: util.Map[Vector4f, util.Map[EnumFacing, IBakedModel]] = new util.HashMap[Vector4f, util.Map[EnumFacing, IBakedModel]]
//
//  private def getOBJModel: OBJModel = {
//    var lightModel: OBJModel = null
//    try {
//      lightModel = OBJLoader.INSTANCE.loadModel(ModelLuxCapatitor.modelLocation).asInstanceOf[OBJModel]
//      lightModel = lightModel.process(ImmutableMap.copyOf(ModelLuxCapatitor.customData)).asInstanceOf[OBJModel]
//    }
//    catch {
//      case e: Exception => {
//        System.out.println("loading model faild!!")
//        lightModel = ModelLoaderRegistry.getMissingModel.asInstanceOf[OBJModel]
//      }
//    }
//    if (lightModel == null) lightModel = ModelLoaderRegistry.getMissingModel.asInstanceOf[OBJModel]
//    return lightModel
//  }
//
//  /**
//    * We need our own because the default set is based on the default=facing north
//    * Our model is default facing up
//    */
//  def getTransform(side: EnumFacing): TRSRTransformation = {
//    var matrix: Matrix4f = null
//    side match {
//      case EnumFacing.DOWN => new TRSRTransformation(ModelRotation.X180_Y0.getMatrix)
////      case EnumFacing.UP =>new TRSRTransformation(TRSRTransformation.identity.getMatrix)
//
//      case NORTH =>
//        matrix = ModelRotation.X90_Y0.getMatrix
//        break //todo: break is not supported
//      case SOUTH =>
//        matrix = ModelRotation.X270_Y0.getMatrix
//        break //todo: break is not supported
//      case WEST =>
//        matrix = ModelRotation.X90_Y270.getMatrix
//        break //todo: break is not supported
//      case EAST =>
//        matrix = ModelRotation.X90_Y90.getMatrix
//        break //todo: break is not supported
//      case _ =>
//        matrix = new Matrix4f
//    }
//    return new TRSRTransformation(matrix)
//  }
//
//  private def bakedLuxCapModel(transformation: TRSRTransformation, facing: EnumFacing, color: Vector4f): IBakedModel = {
//    val model: OBJModel = getOBJModel
//    model.getMatLib.getMaterial(ModelLuxCapatitor.materialName).setColor(color)
//    val bakedModel: IBakedModel = model.bake(transformation, DefaultVertexFormats.BLOCK, new Function[ResourceLocation, TextureAtlasSprite]() {
//      def apply(location: ResourceLocation): TextureAtlasSprite = {
//        val mc: Minecraft = Minecraft.getMinecraft
//        System.out.println("texture location is: " + location.toString)
//        return mc.getTextureMapBlocks.getAtlasSprite(location.toString)
//      }
//    })
//    return bakedModel
//  }
//
//  def getModel(color: Colour, facing: EnumFacing): IBakedModel = {
//    val vector4f: Vector4f = new Vector4f(color.r.toFloat, color.g.toFloat, color.g.toFloat, color.a.toFloat)
//    var luxCapBakedModel: IBakedModel = null
//    try {
//      luxCapBakedModel = registeredModels.get(color).get(facing)
//    }
//    catch {
//      case e: Exception => {
//      }
//    }
//    if (luxCapBakedModel == null) {
//      val tempMap: util.Map[EnumFacing, IBakedModel] = new util.HashMap[EnumFacing, IBakedModel]
//      var tempBaked: IBakedModel = null
//      var i: Int = 0
//      while (i < EnumFacing.VALUES.length) {
//        {
//          val side: EnumFacing = EnumFacing.getFront(i)
//          tempBaked = bakedLuxCapModel(getTransform(side), side, vector4f)
//          if (side eq facing) luxCapBakedModel = tempBaked
//          tempMap.put(side, tempBaked)
//        }
//        {
//          i += 1; i - 1
//        }
//      }
//      registeredModels.put(vector4f, tempMap)
//    }
//    return luxCapBakedModel
//  }
//}

File: src/main/scala/net/machinemuse/powersuits/block/ModelLuxCapatitor.java
package net.machinemuse.powersuits.block;

import com.google.common.collect.ImmutableMap;
import net.machinemuse.numina.geometry.Colour;
import net.machinemuse.powersuits.common.ModularPowersuits;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.block.model.IBakedModel;
import net.minecraft.client.renderer.block.model.ModelRotation;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.client.model.ModelLoaderRegistry;
import net.minecraftforge.client.model.obj.OBJLoader;
import net.minecraftforge.client.model.obj.OBJModel;
import net.minecraftforge.common.model.TRSRTransformation;

import javax.vecmath.Matrix4f;
import javax.vecmath.Vector4f;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by leon on 9/14/16.
 */
public class ModelLuxCapatitor {
    private static ResourceLocation modelLocation = new ResourceLocation(ModularPowersuits.MODID().toLowerCase(), "models/block/lightCore.obj");
//    private static ResourceLocation modelLocation = new ResourceLocation(ModularPowersuits.MODID().toLowerCase(), "models/block/luxcapacitor.obj");

    private static final String materialName = "Material__3.003"; // material of the lens

    private static Map<String, String> customData = new HashMap<String, String>();
    public static final ModelLuxCapatitor instance = new ModelLuxCapatitor();

    private ModelLuxCapatitor(){

    }

    private Map<Integer, Map<EnumFacing, IBakedModel>> registeredModels = new HashMap<>();


    private OBJModel getOBJModel(){
        // this is to fix the texture orientation
        if (!customData.containsKey("flip-v"))
            customData.put("flip-v", "true");

        OBJModel lightModel = null;
        try {
            lightModel = (OBJModel) OBJLoader.INSTANCE.loadModel(modelLocation);
            lightModel = (OBJModel) lightModel.process(ImmutableMap.copyOf(customData));
        } catch (Exception e) {
            System.out.println("loading model faild!!");
            lightModel = (OBJModel) ModelLoaderRegistry.getMissingModel();
        }
        if (lightModel == null)
            lightModel = (OBJModel) ModelLoaderRegistry.getMissingModel();

        return lightModel;
    }

    /**
     * We need our own because the default set is based on the default=facing north
     * Our model is default facing up
     */
    public TRSRTransformation getTransform(EnumFacing side) {
        Matrix4f matrix;

        switch(side.getOpposite()) {
            case DOWN:
                matrix = ModelRotation.X180_Y0.getMatrix();
                break;
            case UP:
                matrix = TRSRTransformation.identity().getMatrix();
                break;
            case NORTH:
                matrix = ModelRotation.X90_Y0.getMatrix();
                break;
            case SOUTH:
                matrix = ModelRotation.X270_Y0.getMatrix();
                break;
            case WEST:
                matrix = ModelRotation.X90_Y270.getMatrix();
                break;
            case EAST:
                matrix = ModelRotation.X90_Y90.getMatrix();
                break;
            default:
                matrix = new Matrix4f();
        }
        return new TRSRTransformation(matrix);
    }


    private IBakedModel bakedLuxCapModel(TRSRTransformation transformation, EnumFacing facing, Vector4f colorVec) {
        OBJModel model = getOBJModel();

        if (model.getMatLib().getMaterialNames().contains(materialName))
            model.getMatLib().getMaterial(materialName).setColor(colorVec);
        else
            System.out.println("Material for LuxCapacitor Lens not found: " + materialName);

        IBakedModel bakedModel = model.bake(transformation, DefaultVertexFormats.ITEM,
                location -> Minecraft.getMinecraft().getTextureMapBlocks().getAtlasSprite(location.toString()));
        return bakedModel;
    }

    public IBakedModel getModel(Colour color, EnumFacing facing) {
        Vector4f colorVec = new Vector4f();

        colorVec.w = (float) color.a;
        colorVec.x = (float) color.r;
        colorVec.y = (float) color.g;
        colorVec.z = (float) color.b;

        IBakedModel luxCapBakedModel = null;
        try {
            luxCapBakedModel = registeredModels.get(color).get(facing);
        } catch (Exception e) {

        }

        if (luxCapBakedModel == null) {
            Map<EnumFacing, IBakedModel> tempMap = new HashMap<>();
            IBakedModel tempBaked;

            for (int i = 0; i < EnumFacing.VALUES.length; i++) {
                EnumFacing side = EnumFacing.getFront(i);
                tempBaked = bakedLuxCapModel(getTransform(side), side, colorVec);
                if (side == facing)
                    luxCapBakedModel = tempBaked;
                tempMap.put(side, tempBaked);
            }
            registeredModels.put(color.getInt(), tempMap);
        }
        return luxCapBakedModel;
    }
}
